File: DMOJ Solutions/ccc06j4.cpp
#include <bits/stdc++.h>
using namespace std;

int vis[10];
vector<int> adj[10];
stack<int> s;
int col = 0;

bool dfs(int u){
	//return true if it finds a cycle
	vis[u] = col;
	for (int v: adj[u]){
		if (vis[v] == vis[u]){
			//in the same iteration we visited v already: there is a cycle
			return true;
		}
		if (!vis[v]){
			if (dfs(v)){
				return true;
			}
		}
		
	}
	//after reaching all the nodes it can reach, add it to the stack
	s.push(u);
	return false;
}

int32_t main(){
	while (true){
		int a,b; cin >> a >> b;
		if (a == 0) break;
		adj[a].emplace_back(b);
	}
	adj[1].emplace_back(7);
	adj[1].emplace_back(4);
	adj[2].emplace_back(1);
	adj[3].emplace_back(4);
	adj[3].emplace_back(5);
	
	for (int i = 7; i >= 1; --i){
		if (!vis[i]){
			++col;
			if (dfs(i)){
				//cycle is found
				cout << "Cannot complete these tasks. Going to bed." << "\n";
				return 0;
			}
		}
	}
	while (!s.empty()){
		cout << s.top() << " ";
		s.pop();
	}
	cout << "\n";
	
	return 0;
}

File: DMOJ Solutions/ccc06j4.py
instructions = [(1,7), (1,4), (2,1), (3,4), (3,5)]
#take input
while True:
    x,y = int(input()), int(input())
    if (x,y) == (0,0):
        break
    else:
        instructions.append((x,y))

stack = []
visited = [0 for i in range(7)]

#sorting the list
def topological_sort(node):
    global stack, instructions, visited
    #set as visited
    visited[node-1] = 1
    #every destination
    temp = [i[1] for i in instructions[len(instructions)-1::-1] if i[0] == node if visited[i[1]-1] == 0]
    #recur
    [topological_sort(n) for n in temp]
    stack.insert(0,node)

    #the function does not need a terminal case because the visited array will eventually be filled


#checking for a cycle
white = set([1,2,3,4,5,6,7])
grey = set()
black = set()
def iscyclic():
    global white, grey, black
    while len(white)>0:
        node = white.pop()
        if dfs(node):
            return True
    return False

def dfs(node):
    global white, grey, black, instructions
    grey.add(node)
    for neighbor in [i[1] for i in instructions if i[0] == node if i[1] not in black]:
        if neighbor in grey:
            return True
        if dfs(neighbor):
            return True
    grey.remove(node)
    return False    

#executing the functions
if iscyclic():
    print("Cannot complete these tasks. Going to bed.")
else:
    for node in [1,2,3,4,5,6,7][6::-1]:
        if visited[node-1] == 0:
            topological_sort(node)
    print(*stack)